File: week 2/strongly_connected/strongly_connected.py
def dfs(adj, x, visited, stack):
    # Mark the current node as visited
    visited[x] = 1
    
    # Recur for all the vertices adjacent to this vertex
    for i in range(len(adj[x])):
        if not visited[adj[x][i]]:
            visited[adj[x][i]] = 1
            dfs(adj, adj[x][i], visited, stack)

    # All vertices reachable from x are processed by now, push x to Stack
    stack.append(x)

def reverseEdges(adj):
    r_adj = [[] for _ in range(len(adj))]
    for i in range(len(adj)):
        for j in range(len(adj[i])):
            r_adj[adj[i][j]].append(i)
    return r_adj

def number_of_strongly_connected_components(adj):
    result = 0
    stack = []

    # Mark all the vertices as not visited (For first DFS)
    visited = [0] * len(adj)

    # Fill vertices in stack according to their finishing times
    for i in range(len(adj)):
        if not visited[i]:
            dfs(adj, i, visited, stack)

    # get the reversed adj list
    r_adj = reverseEdges(adj)

    # Mark all the vertices as not visited (For second DFS)
    visited = [0] * len(adj)

    # Now process all vertices in order defined by Stack
    while stack:
        x = stack.pop()
        if not visited[x]:
            dfs(r_adj, x, visited, [])
            result+=1
    return result

if __name__ == '__main__':
    n, m = map(int, raw_input().split())
    adj = [[] for _ in range(n)]
    for i in range(m):
        a, b = map(int, raw_input().split())
        # adjacency list        
        adj[a - 1].append(b - 1)
    print(number_of_strongly_connected_components(adj))

File: week 2/strongly_connected/StronglyConnected.java
import java.util.ArrayList;
import java.util.Collections;
import java.util.Scanner;
import java.util.Stack;

public class StronglyConnected {
    private static int numberOfStronglyConnectedComponents(ArrayList<Integer>[] adj) {
		int result = 0;
		Stack<Integer> stack = new Stack<>();
		
		// Mark all the vertices as not visited (For first DFS)
        int visited[] = new int[adj.length];
		
		// Fill vertices in stack according to their finishing times
		for(int i = 0; i < adj.length; i++){
			if(visited[i] == 0){
				dfs(adj, i, visited, stack);
			}
		}
		
		// get the reversed adj list
		ArrayList<Integer>[] rAdj = reverseEdges(adj);
		
		// Mark all the vertices as not visited (For second DFS)
		visited = new int[adj.length];
		
	    // Now process all vertices in order defined by Stack
		while (! stack.isEmpty()) {
			// Pop a vertex from stack
            int x = stack.pop();
			
			// get one Strongly connected component of the popped vertex
			if(visited[x] == 0){
				dfs(rAdj, x, visited, new Stack<Integer>());
				result++;
			}
		}
        return result;
    }
	
	private static void dfs(ArrayList<Integer>[] adj, int x, int visited[], Stack<Integer> stack){
		// Mark the current node as visited
		visited[x] = 1;
		
		// Recur for all the vertices adjacent to this vertex
		for(int i = 0; i < adj[x].size(); i++){
			if(visited[adj[x].get(i)] == 0){
				visited[adj[x].get(i)] = 1;
				dfs(adj, adj[x].get(i), visited, stack);
			}	
		}
		
		// All vertices reachable from x are processed by now, push x to Stack
		stack.push(x);
	}
	
	private static ArrayList<Integer>[] reverseEdges(ArrayList<Integer>[] adj){
		ArrayList<Integer>[] rAdj = new ArrayList[adj.length];
		for(int i = 0; i < adj.length; i++){
			rAdj[i] = new ArrayList<Integer>();
		}	
		for(int i = 0; i < adj.length; i++){
			// Recur for all the vertices adjacent to this vertex
			for(int j = 0; j < adj[i].size(); j++){
				rAdj[adj[i].get(j)].add(i);
			}
		}
		return rAdj;
	}

    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);
        int n = scanner.nextInt();
        int m = scanner.nextInt();
        ArrayList<Integer>[] adj = (ArrayList<Integer>[])new ArrayList[n];
        for (int i = 0; i < n; i++) {
            adj[i] = new ArrayList<Integer>();
        }
        for (int i = 0; i < m; i++) {
            int x, y;
            x = scanner.nextInt();
            y = scanner.nextInt();
            adj[x - 1].add(y - 1);
        }
        System.out.println(numberOfStronglyConnectedComponents(adj));
    }
}

